packages/merkle-dag-builder/mod.ts

```diff
@@ -1,313 +1,244 @@
+  },
+  decode(data: Uint8Array): CborValue {
+    return decode(data, (d: Uint8Array) => any.decode(d), undefined).unwrap();
+  },
+};
+
+export type Hash = Uint8Array;
+
+export async function hash(data: BufferSource): Promise<Hash> {
+  return new Uint8Array(await crypto.subtle.digest("SHA-256", data));
 }
+
 // TODO: we need a some way to denote whether the value is a hash
-function isHash(node: LinkValue): node is Uint8Array {
+export function isHash(node: CborValue): node is Hash {
   return node instanceof Uint8Array && node.length === 32;
 }
 
-/**
- * A class that represents a merkle link in a merkle tree / DAG
- */
-export class Link {
-  /** if the value in the link and been modified */
-  public modified: boolean;
-  public hash?: Uint8Array;
-  /** the value stored in the link */
-  public value?: LinkValue;
-  /** on going mutation operations effecting the value of the link */
-  public mutation: Promise<void> | false = false;
-  /**
-   *  A promise that resolve when all the pending write operating started from this Link has finished.
-   * The difference between this and `mutation` is that this is a promise that resolves when all the
-   *  pending write operations have finished and mutation is a promise that resolves when the current mutation on this Link has finished.
-   */
-  public pendingWriteOperation: Promise<void> = Promise.resolve();
-  /**
-   * If no hash is provided the link is considered modified and when merklelized is called
-   * the value will be encoded and hashed to create a new hash. If a hash is provided the value
-   * and a value, it will be assumed that the value corresponds to the hash
-   */
-  constructor(params: {
-    hash?: Uint8Array;
-    value?: LinkValue;
-  }) {
-    assert(params.hash || params.value, "must provide a cid or value");
-    Object.assign(this, params);
-    this.modified = !this.hash;
+function get(obj: CborValue, key: CborKey): CborValue | undefined {
+  if (obj instanceof Map) {
+    return obj.get(key);
+  } else if (
+    typeof obj === "object" && obj !== null &&
+    (typeof key === "number" ||
+      typeof key === "string" ||
+      typeof key === "symbol")
+  ) {
+    return Reflect.get(obj, key);
+  } else {
+    throw new Error(`Cannot get key ${key} from ${obj}`);
   }
+}
 
-  /** acquires a mutation lock, preventing traversal of the link until the lock is resolved */
-  mutationLock(): Omit<PromiseWithResolvers<void>, "promise"> {
-    const { promise, resolve, reject } = Promise.withResolvers<void>();
-    this.mutation = promise;
-    return { resolve, reject };
+function set(obj: CborValue, key: CborKey, value: CborValue): void {
+  if (obj instanceof Map) {
+    obj.set(key, value);
+  } else if (
+    typeof obj === "object" && obj !== null &&
+    (typeof key === "number" ||
+      typeof key === "string" ||
+      typeof key === "symbol")
+  ) {
+    Reflect.set(obj, key, value);
+  } else {
+    throw new Error(`Cannot set key ${key} on ${obj}`);
```

Comment:
this is a bit off topic but i just stumbled over this thought since this function isn't a promise, it might not be obvious to expect this to throw..?

is there something like an "unhandled rejection / exception" linter?